src/psyclone/psyir/nodes/container.py

```diff
@@ -171,21 +171,85 @@ def node_str(self, colour=True):
+        to find the names of the routines that the interface resolves to.
 
-        :param str name: name of the routine to find.
+        If it is not found and the routine is named in an import statement
+        then the search is continued in the named Container. Failing that,
+        all wildcard imports are checked.
 
-        :returns: the PSyIR Routine instance of the subroutine, or None if
-            there is no routine with that name in this container.
-        :rtype: Union[None, psyclone.psyir.nodes.Routine]
+        :param str name: the name of the Routine for which to search.
+        :param bool allow_private: whether the Routine is permitted to have
+            a visibility of PRIVATE.
+
+        :returns: the PSyIR of the named Routine if found, otherwise None.
+        :rtype: :py:class:`psyclone.psyir.nodes.Routine` | NoneType
 
         '''
-        name = name.lower()
-        for routine in self.walk(Routine):
-            if routine.name.lower() == name:
-                return routine
+        rname = name.lower()
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.nodes.routine import Routine
+        from psyclone.psyir.symbols.symbol import Symbol, SymbolError
+        # pylint: enable=import-outside-toplevel
+        for node in self.children:
+            if isinstance(node, Routine) and node.name.lower() == rname:
+                # Check this routine is public
+                routine_sym = self.symbol_table.lookup(node.name)
+                if (allow_private or
+                        routine_sym.visibility == Symbol.Visibility.PUBLIC):
+                    return node
+                # The Container does not contain the expected Routine or the
+                # Routine is not public.
+
+        # Look in the import that names the routine if there is one.
+        table = self.symbol_table
+        try:
+            routine_sym = table.lookup(rname)
+        except KeyError:
+            # Routine does not exist in the SymbolTable.
+            routine_sym = None
+
+        if routine_sym and isinstance(routine_sym, GenericInterfaceSymbol):
+            # The routine is actually an interface to one or more routines.
+            # This should be handled outside this routine (e.g. by using
+            # `resolve_routine` first).
+            return None
+
+        if routine_sym and routine_sym.is_import:
+            child_cntr_sym = routine_sym.interface.container_symbol
+            # Find the definition of the container.
+            container = None
+            try:
+                container = child_cntr_sym.container(local_node=self)
+            except (SymbolError, FileNotFoundError):
+                pass
+            if not container:
+                # TODO #11 log that we didn't find the Container from which
+                # the routine is imported.
+                return None
+            return container.get_routine_psyir(rname)
```

Comment:
Should we pass the same "check_wildcard_imports" value down?

Reply:
We could, but that then means we always fully recurse and that was prohibitive for NEMO where everything is a wildcard import. This way, we only recurse one level and then stop. I've made this more explicit by setting `check_wildcard_imports=False` in this call. To do better we'd need to keep track of the depth.